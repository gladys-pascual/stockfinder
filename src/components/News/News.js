import "./News.scss";
import PropTypes from "prop-types";

const News = ({ news }) => {
  const { category, datetime, headline, image, source, summary, url } = news;

  // Function that  turns the first letter of category uppercase
  // If the category has two words or more, upper case start of each word
  // otherwise capitalize the first letter of the word
  const upperCase = (category) => {
    if (category.includes(" ")) {
      const splitCategory = category.split(" "); //an array
      const upperCasedWords = splitCategory.map((word) => {
        return word[0].toUpperCase() + word.slice(1);
      });
      return upperCasedWords.join(" ");
    } else {
      return category.charAt(0).toUpperCase() + category.slice(1);
    }
  };

  //  For conditional rendering of class name of category
  const categoryClassName = {
    "top news": "top-news",
    business: "business",
  };

  // Converting time given by API to relative time
  function timeDifference(current, previous) {
    const msPerMinute = 60 * 1000;
    const msPerHour = msPerMinute * 60;
    const msPerDay = msPerHour * 24;
    const msPerMonth = msPerDay * 30;
    const msPerYear = msPerDay * 365;
    const elapsed = current - previous;
    if (elapsed < msPerMinute) {
      return Math.round(elapsed / 1000) + " seconds ago";
    } else if (elapsed < msPerHour) {
      return Math.round(elapsed / msPerMinute) + " minutes ago";
    } else if (elapsed < msPerDay) {
      return Math.round(elapsed / msPerHour) + " hours ago";
    } else if (elapsed < msPerMonth && Math.round(elapsed / msPerDay) == 1) {
      return Math.round(elapsed / msPerDay) + " day ago";
    } else if (elapsed < msPerMonth) {
      return Math.round(elapsed / msPerDay) + " days ago";
    } else if (elapsed < msPerYear) {
      return Math.round(elapsed / msPerMonth) + " months ago";
    } else {
      return Math.round(elapsed / msPerYear) + " years ago";
    }
  }
  const currentTime = Date.now();
  const newsTime = timeDifference(currentTime, +`${datetime}000`);

  return (
    <a href={url} target="_blank" rel="noreferrer" className="news-link">
      <section className="news">
        <div className="news-picture">
          <img src={image} alt={headline} />
        </div>
        <div className="news-info">
          <div className="category-source-time">
            <p className={`category ${categoryClassName[category]}`}>
              {upperCase(category)}
            </p>
            <p className="source">{source}</p>
            <p className="time">{newsTime}</p>
          </div>
          <h3 className="headline">{headline}</h3>
          <p className="summary">{summary}</p>
        </div>
      </section>
    </a>
  );
};

News.propTypes = {
  news: PropTypes.object.isRequired,
};

export default News;
