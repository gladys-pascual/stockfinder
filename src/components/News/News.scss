@import "../../assets/variables.scss";

.news-link {
  text-decoration: none;
  color: black;
  display: flex;
  justify-content: center;
  .news {
    margin: 20px 0px 80px 0px;
    display: flex;
    align-items: center;
    border: 1px solid lightgrey;
    border-radius: 4px;
    padding: 8px;
    @media screen and (max-width: 900px) {
      flex-direction: column;
    }
    @media screen and (max-width: 500px) {
      border: none;
    }
    .news-picture {
      width: 40vw;
      padding-right: 20px;
      @media screen and (max-width: 1000px) {
        width: 60vw;
      }
      @media screen and (max-width: 900px) {
        padding-right: 0px;
      }
      @media screen and (max-width: 500px) {
        width: 80vw;
      }
      img {
        width: 100%;
      }
    }
    .news-info {
      @media screen and (max-width: 900px) {
        width: 60vw;
      }
      @media screen and (max-width: 500px) {
        width: 80vw;
      }
      .category-source-time {
        display: flex;
        @media screen and (max-width: 500px) {
          justify-content: center;
        }

        p {
          margin-right: 16px;
        }
        .top-news {
          color: red;
        }
        .business {
          color: blue;
        }
        .source {
          color: rgb(107, 104, 104);
        }
      }
      .headline {
        margin: 0px;
        font-size: 1.5em;
        font-weight: 500;
        font-family: "Roboto", sans-serif;
        letter-spacing: 1px;
        @media screen and (max-width: 500px) {
          text-align: center;
        }
      }
      .summary {
        line-height: 1.5em;
        @media screen and (max-width: 500px) {
          text-align: center;
        }
      }
    }
  }
}
.news-link:hover {
  .news {
    box-shadow: 0 0 10px grey;
    border: 0.5px solid darkgrey;
    @media screen and (max-width: 500px) {
      border: none;
      box-shadow: none;
    }
  }
  .headline {
    color: $theme-color;
  }
}
