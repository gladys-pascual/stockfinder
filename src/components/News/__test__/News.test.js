import { render } from "@testing-library/react";
import News from "../News";

test("render News component as expected", () => {
  const props = {
    news: {
      category: "top news",
      datetime: 1616201855,
      headline:
        "NCAA weight room discrepancy shows 'they don't think these women are worth it,' says Jemele Hill",
      image:
        "https://image.cnbcfm.com/api/v1/image/106441745-1584104407165gettyimages-1137152686.jpeg?v=1584104447",
      source: "CNBC",
      summary:
        "A Stanford University sports performance coach posted photos to Twitter Thursday revealing inequities between the women's and men's weight rooms.",
      url:
        "https://www.cnbc.com/2021/03/19/jemele-hill-ncaa-weight-room-discrepancy-reflects-chronic-gender-inequality.html",
    },
  };

  const { container } = render(<News {...props} />);

  expect(container).toMatchSnapshot();
});
