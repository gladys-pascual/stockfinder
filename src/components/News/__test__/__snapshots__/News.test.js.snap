// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render News component as expected 1`] = `
<div>
  <a
    class="news-link"
    href="https://www.cnbc.com/2021/03/19/jemele-hill-ncaa-weight-room-discrepancy-reflects-chronic-gender-inequality.html"
    rel="noreferrer"
    target="_blank"
  >
    <section
      class="news"
    >
      <div
        class="news-picture"
      >
        <img
          alt="NCAA weight room discrepancy shows 'they don't think these women are worth it,' says Jemele Hill"
          src="https://image.cnbcfm.com/api/v1/image/106441745-1584104407165gettyimages-1137152686.jpeg?v=1584104447"
        />
      </div>
      <div
        class="news-info"
      >
        <div
          class="category-source-time"
        >
          <p
            class="category top-news"
          >
            Top News
          </p>
          <p
            class="source"
          >
            CNBC
          </p>
          <p
            class="time"
          >
            2 days ago
          </p>
        </div>
        <h3
          class="headline"
        >
          NCAA weight room discrepancy shows 'they don't think these women are worth it,' says Jemele Hill
        </h3>
        <p
          class="summary"
        >
          A Stanford University sports performance coach posted photos to Twitter Thursday revealing inequities between the women's and men's weight rooms.
        </p>
      </div>
    </section>
  </a>
</div>
`;
